### PR TITLE
[#75] Accepting wire payments

### LINKED ISSUE #75: Accepting wire payments
## Overview

This is for building out the functionality to support accepting wire payments.

## Additional context

A lot of the functionality was built for another project and just needs to be pulled into this project.

_Including_
- API endpoints to support front-end functionality (creating bank accounts, getting the status, and getting wire instructions) and corresponding types.
- Front-end endpoints and validation.
- Simplify bid flow for wire payments (no need to gather credit car information)
- Checkout flow for initiating creating an account for the wire payment

### PR DESCRIPTION
**Pull request by dallashuggins**
### Changes
- API endpoints to support bank account creation for wire transfers, and for getting the bank account status and instructions for the wire payment. Corresponding endpoints in web.
- If the current bid or pack price is higher than $3000, the wire account flow will be used instead of the payment/card flow. 
  - This includes not requiring card information on the submission of the form, only a checkbox (currently says "I promise to pay", but we should update that) on bid submission, and then initiating creating the wire bank account in Circle. 
  - The user then takes the information from the success page and initiates payment.
  - An email is sent to the user upon successful bank account creation. 
- Feature flag `WIRE_PAYMENT_ENABLED` to enable/disable if the bank account component would be displayed, which is what would handle the bank account creation to initiate the wire transfer flow.

### Examples

Screen Shot 2021-12-03 at 12 49 39 PM

Closes #75 
Progress on #76 
* Last thing: how to handle not awarding a pack before a wire is received

### CODE REVIEW

File: apps/web/src/components/payment-methods-form/index.tsx [hunk 1]
```diff
@@ -8,13 +8,22 @@ import AddMethodsSuccess from './sections/add-methods-success'
...
 export interface MyProfilePaymentMethodsAddProps {
   formErrors?: ExtractError<
-    ReturnType<typeof validatePurchaseForm | typeof validateExpirationDate>
+    ReturnType<
+      | typeof validateBankAccount
+      | typeof validateBidsForm
+      | typeof validateBidsFormForWires
+      | typeof validatePurchaseForm
+      | typeof validateExpirationDate
+    >
```
Comment (dallashuggins) [inline apps/web/src/components/payment-methods-form/index.tsx:26]: The changes in unrelated files are because we're sharing the `formErrors` validation.

File: services/api/src/modules/payments/payments.service.ts [hunk 2]
```diff
@@ -89,6 +91,44 @@ export default class PaymentsService {
...
+    }
+  }
+
+  async getWireTransferInstructions(bankAccountId: string) {
+    // Find bank account in database
+    const foundBankAccount = await PaymentBankAccountModel.query().findById(
+      bankAccountId
+    )
+
+    // If a bank account was found, use the external ID. Otherwise check by passed in ID.
```
Comment (dallashuggins) [inline services/api/src/modules/payments/payments.service.ts:119]:
```suggestion
```

File: services/api/src/modules/payments/payments.service.ts [hunk 3]
```diff
@@ -89,6 +91,44 @@ export default class PaymentsService {
     return cards
   }
 
+  async getBankAccountStatus(bankAccountId: string) {
+    // Find bank account in database
+    const foundBankAccount = await PaymentBankAccountModel.query().findById(
+      bankAccountId
+    )
+
+    // If a bank account was found, use the external ID. Otherwise check by passed in ID.
```
Comment (dallashuggins) [inline services/api/src/modules/payments/payments.service.ts:100]:
```suggestion
```

File: apps/web/src/environment.ts [hunk 4]
```diff
@@ -64,4 +64,19 @@ export const Environment = {
...
+    )
+    switch (isEnabled) {
+      case 'true':
+        return true
+      case 'false':
+        return false
+      default:
+        return isEnabled
+    }
+  },
```
Comment (dallashuggins) [inline apps/web/src/environment.ts:81]: I added this switch because I was not able to get a boolean to return for the env variable, it was always a string.

Comment (smonn) [reply apps/web/src/environment.ts:81]: Right, I guess the `this.config` method is a bit misleading. Can remove the generic typing from it and have it always return a string, and then in cases like this where we need some other type handle it accordingly. The Firebase configs and the currency one are two examples of need something other than a string.

Comment (dallashuggins) [reply apps/web/src/environment.ts:81]: `JSON.parse` seems to always return the right value as well. Was thinking of changing to: ```return typeof JSON.parse(isEnabled) === boolean ? JSON.parse(isEnabled) : false``` Do you see any issue with this?

**Review by smonn — APPROVED** [review]

File: apps/web/src/environment.ts (hunk 4 above)
Comment (dallashuggins) [reply apps/web/src/environment.ts]: Just committed your suggestion, which is nice and clean. Thanks!

Comment (smonn) [reply apps/web/src/environment.ts]: Potential issue with using `JSON.parse` is that if it's an empty string it will throw an error.

File: services/api/src/models/payment-bank-account.model.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import { BaseModel } from './base.model'
+import { UserAccountModel } from './user-account.model'
+
+export class PaymentBankAccountModel extends BaseModel {
+  static tableName = 'PaymentBankAccount'
+  static jsonSchema = PaymentBankAccountSchema
+
+  ownerId!: string
+  externalId!: string
+  status!: PaymentBankAccountStatus | null
```
Comment (smonn) [inline services/api/src/models/payment-bank-account.model.ts:16]: The `status` column in the migration is marked as not nullable. So you'll wanna remove `| null` here or remove the `.notNullable()` from the migration.

```suggestion
  status!: PaymentBankAccountStatus
```

Comment (dallashuggins) [reply services/api/src/models/payment-bank-account.model.ts]: Nice find, thanks!

**Review by jakerainis — APPROVED** [review]

File: services/api/src/migrations/20211116140144_payment_bank_account.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    table.text('description').notNullable()
+    table.uuid('ownerId').notNullable().references('id').inTable('UserAccount')
+    table.timestamp('createdAt').defaultTo(knex.fn.now()).notNullable()
+    table.timestamp('updatedAt').defaultTo(knex.fn.now()).notNullable()
+  })
+  await knex.schema.alterTable('Payment', (table) => {
+    table.uuid('paymentBankId').references('id').inTable('PaymentBankAccount')
+    table.uuid('externalId').nullable().alter()
+  })
+}
```
Comment (dallashuggins) [inline services/api/src/migrations/20211116140144_payment_bank_account.ts:18]: The `Payment` table needs to allow null values for the external ID, so that a payment can be created for the wire transfer before the payment is available via Circle's API. This links the payment as well as the pack to the wire transfer. When the scheduler function is added, we can update the payment with the status and if it fails, remove the claim to the pack.

File: services/api/src/modules/payments/payments.service.ts [hunk 7]
```diff
@@ -315,6 +459,25 @@ export default class PaymentsService {
...
+    })
+    if (!payments) return null
+    // Find payment with matching source ID
+    const sourcePayments = payments.map(
+      (payment) => payment.source.id === sourceId
+    )
+    userInvariant(sourcePayments, 'payments not found', 404)
+    return sourcePayments
+  }
+
```
Comment (dallashuggins) [inline services/api/src/modules/payments/payments.service.ts:480]: For checking for the presence of a matching payment in Circle - will be used in the upcoming scheduler task

File: services/api/src/modules/payments/payments.service.ts [hunk 8]
```diff
@@ -351,20 +588,23 @@ export default class PaymentsService {
...
+            circlePayment,
+            `external payment ${payment.externalId} not found`
+          )
+
+          if (payment.status !== circlePayment.status) {
+            await PaymentModel.query(trx).patchAndFetchById(payment.id, {
+              status: circlePayment.status,
+            })
+            updatedPayments++
+          }
```
Comment (dallashuggins) [inline services/api/src/modules/payments/payments.service.ts:605]: Since `externalId` can be null, this conditional was needed.

File: services/api/src/modules/payments/payments.service.ts [hunk 9]
```diff
@@ -149,6 +193,92 @@ export default class PaymentsService {
...
+
+    // Create new payment in database
+    const newPayment = await PaymentModel.query(trx).insert({
+      externalId: null,
+      payerId: user.id,
+      packId: packId,
+      paymentBankId: newBankAccount.id,
+      paymentCardId: null,
+      status: PaymentStatus.Pending,
+    })
```
Comment (dallashuggins) [inline services/api/src/modules/payments/payments.service.ts:251]: Adding a payment now during the bank account creation.

File: apps/web/src/components/bank-account-form/sections/bank-account-form.tsx [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+                    : '',
+                country:
+                  formErrors && 'country' in formErrors
+                    ? (formErrors.country as string)
+                    : '',
+                zipCode:
+                  formErrors && 'zipCode' in formErrors
+                    ? (formErrors.zipCode as string)
+                    : '',
+              }}
```
Comment (smonn) [inline apps/web/src/components/bank-account-form/sections/bank-account-form.tsx:152]: Could we align the types here and in other places (for `formErrors`) to avoid the repeated checks and casting?

File: apps/web/src/components/bank-account-form/sections/bank-account-success.tsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+  const isActiveAuction =
+    release.type === PackType.Auction &&
+    isAfterNow(new Date(release.auctionUntil as string))
+
+  const handleReturnToListing = useCallback(() => {
+    const path = urls.release.replace(':packSlug', release.slug)
+    if (typeof window !== 'undefined') {
+      window.location.assign(new URL(path, window.location.origin).href)
+    }
```
Comment (smonn) [inline apps/web/src/components/bank-account-form/sections/bank-account-success.tsx:36]: Is using `router.push(path)` not viable here? (where `const router = useRouter()`)

Comment (dallashuggins) [reply apps/web/src/components/bank-account-form/sections/bank-account-success.tsx:36]: Good call! Changed in the other branch with [fd3d2bd](https://github.com/deptagency/algomart/pull/123/commits/fd3d2bd79ade44c23c4d081fd5b9d37e269a3de5)

File: apps/web/src/components/bank-account-form/sections/bank-account-form.tsx (hunk 10 above)
Comment (dallashuggins) [reply apps/web/src/components/bank-account-form/sections/bank-account-form.tsx:152]: Did you have an idea how to do this? I can look into it more, definitely would be ideal, but since we're group the different validations this is the only way I know to make it not throw an error saying the property doesn't exist.

**Review by smonn — APPROVED** [review]

**Comment by afraser** [comment]
Small thing but the "success" messaging might give people the wrong idea (ie: that they're done and that the following text is just informational). Should we remove that?

**Comment by dallashuggins** [comment]
> Small thing but the "success" messaging might give people the wrong idea (ie: that they're done and that the following text is just informational). Should we remove that?

@afraser Good point! Since this is feature flagged I could add in a future PR if we wanted to get this in, or do similar to the other PR and merge into this PR, but I think I'll update the success screen and general style according to the design Tom worked on.

File: apps/web/src/utils/purchase-validation.ts [hunk 12]
```diff
@@ -8,12 +8,17 @@ import {
...
   object,
+  oneOf,
   required,
   string,
 } from 'validator-fns'
 
 import { formatCurrency } from '@/utils/format-currency'
 
+// Maximum bid for card payments as integer
+// eslint-disable-next-line unicorn/numeric-separators-style
```
Comment (dallashuggins) [inline apps/web/src/utils/purchase-validation.ts:19]: Throws the below error with `3_000_00` and this is an integer, so the disable line is needed. 
```
Invalid group length in numeric value.eslintunicorn/numeric-separators-style
```

**Comment by dallashuggins** [comment]
@afraser I did end up removing "Success" in this PR and making the style closer to the designs Tom made.

File: apps/web/src/components/checkbox.tsx [hunk 13]
```diff
@@ -2,30 +2,33 @@ import clsx from 'clsx'
...
+            'bg-gray-200': readOnly || disabled,
+          },
+          className
+        )}
+        checked={checked}
+        disabled={disabled}
+        readOnly={readOnly}
+        {...props}
+      />
+      {label && <span className="pl-2">{label}</span>}
```
Comment (afraser) [inline apps/web/src/components/checkbox.tsx:31]: nice!

File: apps/web/src/components/bank-account-form/sections/bank-account-success.tsx [hunk 14]
```diff
@@ -38,28 +38,19 @@ export default function BankAccountSuccess({
   return (
     <div className={css.root}>
       <CheckCircleIcon className={css.icon} height="48" width="48" />
-      {release.type === PackType.Auction && (
+      {release.type === PackType.Auction && isActiveAuction && (
```
Comment (afraser) [inline apps/web/src/components/bank-account-form/sections/bank-account-success.tsx:45]: nice!

**Review by afraser — APPROVED** [review]

File: apps/web/src/components/bid-activity/bid-activity.tsx [hunk 15]
```diff
@@ -113,7 +112,7 @@ export default function BidActivity({
                 )}
               </BidActivityDetails>
               {meetsReservePrice &&
-                (!isFollowingBid || followingBidDoesNotMeetReservePrice) && (
+                (!followingBid || followingBidDoesNotMeetReservePrice) && (
```
Comment (dallashuggins) [inline apps/web/src/components/bid-activity/bid-activity.tsx:115]: The "Reserve price met" activity was never being displayed, since the reserve price (price from the CMS) is an integer, and we were converting to an integer again and therefore the price was always higher.
